core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@junrao: Yeah that is the right process. The speedup is in step (a). In step (a) there will be 3000 requests queued before ControlledShutdownResponse in the request queue from controller to broker. It takes much more time to process 3000 requests of 1 partition each, instead of 3 requests of 1000 partitions each if broker is busy processing requests from many other queues. The more active connections that the broker has with clients and brokers, the more time it takes for broker to process any single request from any given queue, because broker takes requests from all non-empty queues in a round-robin fashion. Given that there is only limited number of physical cores on the broker, the expected time to finish processing that 3000 the requests (so that broker can finally process ControlledShutdownResponse and actually shutdown) is proportional to `number_of_requests_in_the_queue * number_of_non_empty_queues`. The number of non-empty queue should be equivalent to number of active channels with brokers or clients.

Yeah we can optimize the RequestSendThread as you said to reduce the number of requests sent by controller for controlled shutdown. I am not 100% sure if it can achieve similar or better performance as the approach in this patch. Because RequestSendThread may have already sent multiple requests during the time when that 3000 requests are generated and put into the queue. I can certainly test it and compare the two approaches.